### PR TITLE
Fix a crash problem caused by GetActorHandle in ActorManager

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

ActorManager::GetActorHandle now returns a reference of unqiue_ptr<ActorHandle>. This will cause a crash in an edge case:

-  Thread 1 calls GetActorHandle and gets a unique_ptr<ActorHandle> &.
-  Thread 2 calls AddActorHandle and triggers actor_handles_ flat map to resize.
-  Then the unique_ptr<ActorHandle> & in step 1 becomes invalid.

This PR fixes this issue by using shared_ptr.

Note, this bug only happens in stressful scenarios and takes long time to reproduce, so I didn't add unit test. But it is already verified in our production environment.

This PR also changes `CoreWorker::GetActorHandle` to return a `shared_ptr` to Python/Java, instead of a raw pointer. Using a raw pointer is okay for now because we don't delete actor handles. But if we do in the future, there will be a wild pointer issue.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -1321,7 +1322,11 @@ cdef class CoreWorker:
         CCoreWorkerProcess.GetCoreWorker().RemoveActorHandleReference(
             c_actor_id)
 
-    cdef make_actor_handle(self, const CActorHandle *c_actor_handle):
+    cdef make_actor_handle(self, ActorHandleSharedPtr actor_handle_shared_ptr):
```
Comment (edoakes) [inline python/ray/_raylet.pyx:1325]: this function should probably take a shared pointer ref to avoid incrementing the shared ptr

**Comment by rkooo567** [comment]
I remember there were some performance regression when I tried using shared ptr. Can you run the microbenchmark to see the impact? (it is not that I disagree with the change, but I'd just like to see it)

File: python/ray/_raylet.pyx (hunk 1 above)
Comment (raulchen) [reply python/ray/_raylet.pyx:1325]: Cython doesn't support reference. Copying a `shared_ptr` is cheap. So I think it's fine.

File: python/ray/_raylet.pyx [hunk 2]
```diff
@@ -1321,7 +1322,11 @@ cdef class CoreWorker:
...
+        cdef:
+            # NOTE: This raw pointer is only valid in this function, while the
+            # shared_ptr is valid. Do not store it anywhere.
+            const CActorHandle *c_actor_handle = actor_handle_shared_ptr.get()
```
[the remark this answers is not in the document]
Comment (raulchen) [reply python/ray/_raylet.pyx:1329]: Not particular reason. I was just thinking using too many `deref`s would be a bit too verbose. But I'm okay with that as well.

**Comment by raulchen** [comment]
@rkooo567  There are small differences. But look more like normal fluctuations. Theoretically, this change shouldn't have noticeable perf impact.

Without this PR:
```
2021-01-05 11:37:14,611 INFO services.py:1171 -- View the Ray dashboard at http://127.0.0.1:8265
single client get calls per second 15026.95 +- 146.65
single client put calls per second 10872.65 +- 137.85
multi client put calls per second 43057.72 +- 1004.63                                                                          
2021-01-05 11:37:45,356 INFO services.py:1171 -- View the Ray dashboard at http://127.0.0.1:8265
single client get calls (Plasma Store) per second 5107.88 +- 39.19
single client put calls (Plasma Store) per second 2637.34 +- 121.42
multi client put calls (Plasma Store) per second 4317.33 +- 120.29
single client put gigabytes per second 18.79 +- 0.43
multi client put gigabytes per second 12.15 +- 0.92
single client tasks sync per second 1006.16 +- 10.18
single client tasks async per second 7133.38 +- 189.81
multi client tasks async per second 8549.28 +- 302.23
1:1 actor calls sync per second 1672.91 +- 39.49
1:1 actor calls async per second 4299.04 +- 19.17
1:1 actor calls concurrent per second 4514.88 +- 113.25
1:n actor calls async per second 7283.19 +- 384.86
n:n actor calls async per second 9519.81 +- 533.55
n:n actor calls with arg async per second 4310.6 +- 77.02
1:1 async-actor calls sync per second 1095.01 +- 33.56
1:1 async-actor calls async per second 2137.72 +- 31.71
1:1 async-actor calls with args async per second 1369.75 +- 105.79
1:n async-actor calls async per second 5053.99 +- 113.67
```
With this PR:
```
2021-01-05 11:51:13,466 INFO services.py:1171 -- View the Ray dashboard at http://127.0.0.1:8265
single client get calls per second 13948.63 +- 375.41
single client put calls per second 10641.19 +- 362.68
multi client put calls per second 45601.29 +- 851.94
2021-01-05 11:51:44,055 INFO services.py:1171 -- View the Ray dashboard at http://127.0.0.1:8265
single client get calls (Plasma Store) per second 5034.5 +- 65.55
single client put calls (Plasma Store) per second 2691.77 +- 137.73
multi client put calls (Plasma Store) per second 4511.46 +- 51.54
single client put gigabytes per second 18.57 +- 0.3
multi client put gigabytes per second 11.12 +- 0.58
single client tasks sync per second 1007.37 +- 22.17
single client tasks async per second 7036.4 +- 230.03
multi client tasks async per second 8879.48 +- 160.02
1:1 actor calls sync per second 1668.2 +- 34.57
1:1 actor calls async per second 4220.3 +- 74.43
1:1 actor calls concurrent per second 4477.59 +- 72.97
1:n actor calls async per second 7358.65 +- 225.1
n:n actor calls async per second 9809.32 +- 352.21
n:n actor calls with arg async per second 4609.57 +- 137.63
1:1 async-actor calls sync per second 1103.12 +- 18.85
1:1 async-actor calls async per second 2101.06 +- 26.36
1:1 async-actor calls with args async per second 1389.54 +- 30.64
1:n async-actor calls async per second 5101.26 +- 58.1
```

**Comment by raulchen** [comment]
Test failures should be unrelated.

**Review by kfstorm — APPROVED** [review]

**Pull request merged**